### ISSUE TITLE
What social login providers should we support?

### ISSUE THREAD

**Issue by esjewett** [humanitiesplusdesign/fibra#146]

**Comment by jiemakel** [comment]
Found another one: https://github.com/sahat/satellizer . Here, we'd get all of: Google, Facebook, LinkedIn, Twitter, Instagram, GitHub, Bitbucket, Yahoo, Twitch, Microsoft (Windows Live) in a package.

**Comment by esjewett** [comment]
I think that only handles the first part of the flow. Exchanging the temporary code for the access token seems to be expected to happen server-side. But still may be useful to use to implement using this library to avoid maintaining custom stuff. I'll look into it further. Thanks!

**Comment by jiemakel** [comment]
Hmm, you're right. So this does need a bit more thinking through. Though, I think the [angularjs-social-login](https://github.com/heresy/angularjs-social-login) is fully client side, so that would give use Google, Facebook and LinkedIn. satellizer itself also seems to support the implicit grant flow, which is aimed at client side login, but there's scant documentation on that. I also found [oauth-ng](https://github.com/angularjs-oauth/oauth-ng) which seems to be a generic implicit flow authenticator.

Finally, there's also [OpenID Connect](http://openid.net/connect/), which is supported by e.g. [Google and Yahoo](http://openid.net/get-an-openid/) and has a [Facebook bridge](http://facebook-openid.appspot.com/). Didn't find any killer Angular library for that one though. There is [this](https://github.com/JamesRandall/AngularJS-OAuth2) for Angular2 and [this](https://github.com/manfredsteyer/angular-oauth-oidc) seems to be an example of OAuth2/OIDC for Angular1.

**Comment by esjewett** [comment]
Yeah, there are a bunch of options. 5 years ago I was an OAuth nerd, but have thankfully left that well behind. This is bringing back bad memories :-) I think the auth flow without client secret works for a bunch of platforms but unfortunately not Github as they don't have CORS enabled on their access-token endpoint, so I have to use a relay running on a Heroku dyno, at least temporarily :-( But I actually think Satellizer will be really helpful in handling the initial part of the authorization flow generically and then we can handle each service in whatever way is necessary as far as exchanging the code for an access token - for services that support more modern client-side auth flows than Github, that part is very simple, as demonstrated by the angularjs-social-login library.

For now, I think I'm just going to leave it as is until we determine what login providers we actually need to support. If Github is enough, then that works already, though I need to fix up the UI a little. If we need more, then will switch over to Satellizer, likely.

**Comment by esjewett** [comment]
For now we'll stick with Github. Longer term we need to add more providers and possibly include institutional providers. Probably will require a server-side component.

**Comment by jiemakel** [comment]
I suggest we also add Google and Facebook right away if they come relatively easily, as not all of our test users probably have GitHub accounts (that is, if, as now, we require a login to even dabble with the app)

**Comment by cncoleman** [comment]
Facebook is evil. How about Google and Twitter? We already talked about OpenID, right? It's complicated?

**Comment by jiemakel** [comment]
Just Google then for a start?

**Comment by esjewett** [comment]
I think OpenID use is pretty rare now. I haven't see an opportunity to use mine for a long time, though I think Google uses it behind the scenes. Login with Twitter/Google should be doable.

Google will be quite easy and doesn't require a server-side component that I see: https://developers.google.com/identity/sign-in/web/sign-in

Twitter is more of a pain and will require a server-side component like Github because their access_token endpoint doesn't support CORS: https://dev.twitter.com/web/sign-in/implementing

Still thinking it may make the most sense to integrate Satellizer and use that for the initial redirect flow. But we just need to prioritize when we do this based on how important it is.

**Comment by esjewett** [comment]
Oh, Facebook is as easy as Google, and Zuckerberg seems to be seeing the light a little bit, so who knows: https://developers.facebook.com/docs/facebook-login/web

Might make sense to do a quick survey in the Slack and direct people to it via email. Asking what IDs they have and what are preferred.

**Comment by jiemakel** [comment]
I'm also already using a cors proxy on ldf.fi to access the Getty endpoint, if that would help. On the other hand, the easiest approach for now would be to just disable the need for being authenticated until we actually use it for anything.

**Comment by esjewett** [comment]
Yes, true. The reason I enabled it was because I wanted to get it tested and didn't want to handle the logged out case in the header bar unnecessarily. But happy to just disable the automatic login and add a "Login" link in the header bar when not logged in.
